File: app/services/billing_service.py
"""
Billing service — orquestra o fluxo de signup com trial:
  1. Cria/recupera customer no Asaas
  2. Cria subscription recorrente com trialEndDate
  3. Persiste AsaasCustomer + Subscription + Invoice no banco
  4. Ativa user (is_active=True, plan_type=plan, status=TRIALING)

Grace period:
  Quando um pagamento atrasa, o usuário recebe 48h antes de ser desativado.
  O job `expire_grace_periods` roda a cada hora verificando assinaturas expiradas.
"""
import logging
import uuid
from datetime import date, timedelta, datetime, timezone

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.config import get_settings
from app.models.billing import AsaasCustomer, Subscription, Invoice, SubStatus, WebhookEvent
from app.models.user import User
from app.services import asaas_client
from app.services import chatwoot_client

logger = logging.getLogger(__name__)
settings = get_settings()

GRACE_HOURS: int = settings.GRACE_HOURS  # configurável via .env (padrão 48h)


async def start_trial(
    db: AsyncSession,
    user: User,
    plan: str,
    card_data: dict,
    card_holder_info: dict,
) -> dict:
    """
    Ponto de entrada chamado pelo /auth/signup.

    Retorna:
        {
            "subscription_id": str,
            "trial_end_date": str (YYYY-MM-DD),
            "plan": str,
        }

    Lança httpx.HTTPStatusError se o Asaas rejeitar o cartão.
    """
    plan = plan.upper()

    # ── 1. Asaas customer ─────────────────────────────────────────────────
    # Verifica se já existe registro local
    existing = await db.execute(
        select(AsaasCustomer).where(AsaasCustomer.user_id == user.id)
    )
    ac = existing.scalars().first()

    if ac:
        asaas_customer_id = ac.asaas_customer_id
    else:
        customer = await asaas_client.get_or_create_customer(
            name=user.full_name or "",
            phone=user.wa_id,
            email=user.email,
        )
        asaas_customer_id = customer["id"]
        db.add(AsaasCustomer(user_id=user.id, asaas_customer_id=asaas_customer_id))
        await db.flush()

    # ── 2. Asaas subscription ─────────────────────────────────────────────
    subscription_data = await asaas_client.create_subscription(
        customer_id=asaas_customer_id,
        plan=plan,
        card_data=card_data,
        card_holder_info=card_holder_info,
        trial_days=settings.TRIAL_DAYS,
    )

    asaas_subscription_id = subscription_data["id"]
    trial_end = date.today() + timedelta(days=settings.TRIAL_DAYS)

    # ── 3. Persist subscription ───────────────────────────────────────────
    sub = Subscription(
        user_id=user.id,
        asaas_subscription_id=asaas_subscription_id,
        status=SubStatus.TRIALING,
        plan_name=plan,
        next_due_date=trial_end,
    )
    db.add(sub)
    await db.flush()

    # ── 4. Activate user ──────────────────────────────────────────────────
    user.is_active = True
    user.plan_type = plan
    db.add(user)
    await db.commit()

    return {
        "subscription_id": asaas_subscription_id,
        "trial_end_date": trial_end.isoformat(),
        "plan": plan,
    }


async def handle_payment_confirmed(db: AsyncSession, payment_id: str, subscription_id: str | None) -> None:
    """
    Webhook PAYMENT_CONFIRMED / PAYMENT_RECEIVED:
    - Registra invoice como paga
    - Garante user ativo
    """
    if not subscription_id:
        return

    sub_result = await db.execute(
        select(Subscription).where(Subscription.asaas_subscription_id == subscription_id)
    )
    sub = sub_result.scalars().first()
    if not sub:
        return

    # Upsert invoice
    inv_result = await db.execute(
        select(Invoice).where(Invoice.asaas_payment_id == payment_id)
    )
    inv = inv_result.scalars().first()
    if not inv:
        inv = Invoice(
            user_id=sub.user_id,
            subscription_id=sub.id,
            asaas_payment_id=payment_id,
            amount=asaas_client.PLAN_VALUES.get(sub.plan_name or "", 0),
            status="CONFIRMED",
        )
        db.add(inv)
    else:
        inv.status = "CONFIRMED"

    # Ensure user is active and subscription ACTIVE
    sub.status = SubStatus.ACTIVE
    user_result = await db.execute(select(User).where(User.id == sub.user_id))
    user = user_result.scalars().first()
    if user:
        user.is_active = True

    await db.commit()


async def handle_payment_overdue(db: AsyncSession, payment_id: str, subscription_id: str | None) -> None:
    """
    Webhook PAYMENT_OVERDUE / PAYMENT_CREDIT_CARD_CAPTURE_REFUSED:
    - Upsert invoice como OVERDUE (cria se não existir)
    - Marca subscription como PAST_DUE com grace_expires_at = agora + 48h
    - Notifica o usuário via WhatsApp
    - NÃO desativa imediatamente — o job expire_grace_periods faz isso
    """
    if not subscription_id:
        return

    sub_result = await db.execute(
        select(Subscription).where(Subscription.asaas_subscription_id == subscription_id)
    )
    sub = sub_result.scalars().first()
    if not sub:
        return

    # Define grace period apenas se ainda não estiver em PAST_DUE (evita resetar o timer)
    if sub.status != SubStatus.PAST_DUE:
        sub.status = SubStatus.PAST_DUE
        sub.grace_expires_at = datetime.now(timezone.utc) + timedelta(hours=GRACE_HOURS)

        # Notifica o usuário via WhatsApp
        user_result = await db.execute(select(User).where(User.id == sub.user_id))
        user = user_result.scalars().first()
        if user:
            await _notify_payment_overdue(user.wa_id, GRACE_HOURS)

    inv_result = await db.execute(
        select(Invoice).where(Invoice.asaas_payment_id == payment_id)
    )
    inv = inv_result.scalars().first()
    if inv:
        inv.status = "OVERDUE"
    elif payment_id:
        # Pagamento ainda não existia localmente — cria o registro
        inv = Invoice(
            user_id=sub.user_id,
            subscription_id=sub.id,
            asaas_payment_id=payment_id,
            amount=asaas_client.PLAN_VALUES.get(sub.plan_name or "", 0),
            status="OVERDUE",
        )
        db.add(inv)

    await db.commit()


async def _notify_payment_overdue(wa_id: str, grace_hours: int) -> None:
    """Envia mensagem de pagamento atrasado via WhatsApp."""
    message = (
        f"⚠️ *Koalla — Pagamento pendente*\n\n"
        f"Identificamos um problema com a cobrança da sua assinatura.\n\n"
        f"Você tem *{grace_hours} horas* para regularizar antes que o acesso seja suspenso.\n\n"
        f"Se precisar de ajuda, é só responder aqui. 🐨"
    )
    try:
        await chatwoot_client.send_message_to_phone(wa_id, message)
    except Exception as exc:
        logger.warning("Falha ao notificar usuário %s sobre pagamento atrasado: %s", wa_id, exc)


async def expire_grace_periods(db: AsyncSession) -> int:
    """
    Job periódico — desativa usuários cujo grace period expirou.
    Chamado pelo APScheduler a cada hora.
    Retorna o número de usuários desativados.
    """
    now = datetime.now(timezone.utc)
    result = await db.execute(
        select(Subscription).where(
            Subscription.status == SubStatus.PAST_DUE,
            Subscription.grace_expires_at <= now,
        )
    )
    expired_subs = result.scalars().all()

    deactivated = 0
    for sub in expired_subs:
        sub.status = SubStatus.CANCELED

        user_result = await db.execute(select(User).where(User.id == sub.user_id))
        user = user_result.scalars().first()
        if user and user.is_active:
            user.is_active = False
            deactivated += 1
            logger.info("Usuário %s desativado por grace period expirado", user.wa_id)
            # Notifica encerramento
            await _notify_access_revoked(user.wa_id)

    if expired_subs:
        await db.commit()

    return deactivated


async def _notify_access_revoked(wa_id: str) -> None:
    """Notifica o usuário que o acesso foi suspenso."""
    message = (
        "🔒 *Koalla — Acesso suspenso*\n\n"
        "Infelizmente não conseguimos processar o pagamento da sua assinatura "
        "e seu acesso foi suspenso.\n\n"
        "Para reativar, entre em contato com o suporte. 🐨"
    )
    try:
        await chatwoot_client.send_message_to_phone(wa_id, message)
    except Exception as exc:
        logger.warning("Falha ao notificar usuário %s sobre suspensão: %s", wa_id, exc)


async def handle_subscription_deleted(db: AsyncSession, subscription_id: str) -> None:
    """
    Webhook SUBSCRIPTION_DELETED / SUBSCRIPTION_INACTIVATED:
    - Marca subscription como CANCELED
    - Desativa user
    """
    sub_result = await db.execute(
        select(Subscription).where(Subscription.asaas_subscription_id == subscription_id)
    )
    sub = sub_result.scalars().first()
    if not sub:
        return

    sub.status = SubStatus.CANCELED

    user_result = await db.execute(select(User).where(User.id == sub.user_id))
    user = user_result.scalars().first()
    if user:
        user.is_active = False

    await db.commit()


# ── Idempotência ───────────────────────────────────────────────────────────

async def is_event_already_processed(db: AsyncSession, event_id: str) -> bool:
    """
    Verifica se o event_id já foi processado.
    O Asaas envia webhooks com entrega "at least once" — duplicatas devem ser ignoradas.
    """
    result = await db.execute(
        select(WebhookEvent).where(WebhookEvent.event_id == event_id)
    )
    return result.scalars().first() is not None


async def mark_event_processed(db: AsyncSession, event_id: str, event_type: str) -> None:
    """
    Registra o event_id como processado.
    Chamado após o processamento bem-sucedido de qualquer evento.
    """
    db.add(WebhookEvent(event_id=event_id, event_type=event_type))
    await db.commit()

File: src/main/kotlin/ai/koalla/core/service/BillingService.kt
package ai.koalla.core.service

import ai.koalla.core.client.AsaasClient
import ai.koalla.core.client.ChatwootClient
import ai.koalla.core.config.KoallaProperties
import ai.koalla.core.dto.CardData
import ai.koalla.core.dto.CardHolderInfo
import ai.koalla.core.entity.*
import ai.koalla.core.repository.*
import kotlinx.coroutines.runBlocking
import org.slf4j.LoggerFactory
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDate
import java.time.OffsetDateTime
import java.util.UUID

/**
 * Billing service — orchestrates signup with trial:
 *   1. Creates/retrieves customer in Asaas
 *   2. Creates recurring subscription with trialEndDate
 *   3. Persists AsaasCustomer + Subscription + Invoice in database
 *   4. Activates user (is_active=True, plan_type=plan, status=TRIALING)
 *
 * Grace period:
 *   When a payment is late, the user gets 48h before being deactivated.
 *   The job `expireGracePeriods` runs every hour checking expired subscriptions.
 */
@Service
class BillingService(
    private val userRepository: UserRepository,
    private val asaasCustomerRepository: AsaasCustomerRepository,
    private val subscriptionRepository: SubscriptionRepository,
    private val invoiceRepository: InvoiceRepository,
    private val webhookEventRepository: WebhookEventRepository,
    private val asaasClient: AsaasClient,
    private val chatwootClient: ChatwootClient,
    private val props: KoallaProperties
) {
    private val logger = LoggerFactory.getLogger(javaClass)

    /**
     * Entry point called by /auth/signup.
     */
    @Transactional
    suspend fun startTrial(
        user: User,
        plan: String,
        cardData: CardData,
        cardHolderInfo: CardHolderInfo
    ): TrialResult {
        val planUpper = plan.uppercase()

        // 1. Asaas customer
        val existingCustomer = asaasCustomerRepository.findByUserId(user.id!!)
        val asaasCustomerId: String

        if (existingCustomer != null) {
            asaasCustomerId = existingCustomer.asaasCustomerId
        } else {
            val customer = asaasClient.getOrCreateCustomer(
                name = user.fullName ?: "",
                phone = user.waId,
                email = user.email
            )
            asaasCustomerId = customer["id"] as String
            asaasCustomerRepository.save(
                AsaasCustomer(
                    userId = user.id!!,
                    asaasCustomerId = asaasCustomerId
                )
            )
        }

        // 2. Asaas subscription
        val subscriptionData = asaasClient.createSubscription(
            customerId = asaasCustomerId,
            plan = planUpper,
            cardData = cardData,
            cardHolderInfo = cardHolderInfo,
            trialDays = props.trialDays
        )

        val asaasSubscriptionId = subscriptionData["id"] as String
        val trialEnd = LocalDate.now().plusDays(props.trialDays.toLong())

        // 3. Persist subscription
        val subscription = Subscription(
            userId = user.id!!,
            asaasSubscriptionId = asaasSubscriptionId,
            status = SubStatus.TRIALING,
            planName = planUpper,
            nextDueDate = trialEnd
        )
        subscriptionRepository.save(subscription)

        // 4. Activate user
        user.isActive = true
        user.planType = planUpper
        userRepository.save(user)

        return TrialResult(
            subscriptionId = asaasSubscriptionId,
            trialEndDate = trialEnd.toString(),
            plan = planUpper
        )
    }

    /**
     * Webhook PAYMENT_CONFIRMED / PAYMENT_RECEIVED:
     * - Records invoice as paid
     * - Ensures user is active
     */
    @Transactional
    fun handlePaymentConfirmed(paymentId: String, subscriptionId: String?) {
        if (subscriptionId == null) return

        val subscription = subscriptionRepository.findByAsaasSubscriptionId(subscriptionId) ?: return

        // Upsert invoice
        var invoice = invoiceRepository.findByAsaasPaymentId(paymentId)
        if (invoice == null) {
            invoice = Invoice(
                userId = subscription.userId,
                subscriptionId = subscription.id,
                asaasPaymentId = paymentId,
                amount = props.plans.getValue(subscription.planName ?: ""),
                status = "CONFIRMED"
            )
        } else {
            invoice.status = "CONFIRMED"
        }
        invoiceRepository.save(invoice)

        // Ensure user is active and subscription ACTIVE
        subscription.status = SubStatus.ACTIVE
        subscription.graceExpiresAt = null
        subscriptionRepository.save(subscription)

        val user = userRepository.findById(subscription.userId).orElse(null)
        if (user != null && !user.isActive) {
            user.isActive = true
            userRepository.save(user)
        }
    }

    /**
     * Webhook PAYMENT_OVERDUE / PAYMENT_CREDIT_CARD_CAPTURE_REFUSED:
     * - Upsert invoice as OVERDUE
     * - Marks subscription as PAST_DUE with grace_expires_at = now + 48h
     * - Notifies user via WhatsApp
     * - Does NOT deactivate immediately — the job expireGracePeriods does that
     */
    @Transactional
    fun handlePaymentOverdue(paymentId: String, subscriptionId: String?) {
        if (subscriptionId == null) return

        val subscription = subscriptionRepository.findByAsaasSubscriptionId(subscriptionId) ?: return

        // Set grace period only if not already PAST_DUE (avoid resetting timer)
        if (subscription.status != SubStatus.PAST_DUE) {
            subscription.status = SubStatus.PAST_DUE
            subscription.graceExpiresAt = OffsetDateTime.now().plusHours(props.graceHours.toLong())
            subscriptionRepository.save(subscription)

            // Notify user via WhatsApp
            val user = userRepository.findById(subscription.userId).orElse(null)
            if (user != null) {
                runBlocking {
                    notifyPaymentOverdue(user.waId)
                }
            }
        }

        // Upsert invoice
        var invoice = invoiceRepository.findByAsaasPaymentId(paymentId)
        if (invoice != null) {
            invoice.status = "OVERDUE"
        } else {
            invoice = Invoice(
                userId = subscription.userId,
                subscriptionId = subscription.id,
                asaasPaymentId = paymentId,
                amount = props.plans.getValue(subscription.planName ?: ""),
                status = "OVERDUE"
            )
        }
        invoiceRepository.save(invoice)
    }

    private suspend fun notifyPaymentOverdue(waId: String) {
        val message = """
            ⚠️ *Koalla — Pagamento pendente*
            
            Identificamos um problema com a cobrança da sua assinatura.
            
            Você tem *${props.graceHours} horas* para regularizar antes que o acesso seja suspenso.
            
            Se precisar de ajuda, é só responder aqui. 🐨
        """.trimIndent()

        try {
            chatwootClient.sendMessageToPhone(waId, message)
        } catch (e: Exception) {
            logger.warn("Failed to notify user $waId about overdue payment: ${e.message}")
        }
    }

    /**
     * Periodic job — deactivates users whose grace period has expired.
     * Called by Spring Scheduler every hour.
     */
    @Scheduled(fixedRate = 3600000) // 1 hour in milliseconds
    @Transactional
    fun expireGracePeriods(): Int {
        val now = OffsetDateTime.now()
        val expiredSubs = subscriptionRepository.findExpiredGracePeriods(now)

        var deactivated = 0
        for (sub in expiredSubs) {
            sub.status = SubStatus.CANCELED
            subscriptionRepository.save(sub)

            val user = userRepository.findById(sub.userId).orElse(null)
            if (user != null && user.isActive) {
                user.isActive = false
                userRepository.save(user)
                deactivated++
                logger.info("User ${user.waId} deactivated due to expired grace period")

                runBlocking {
                    notifyAccessRevoked(user.waId)
                }
            }
        }

        if (deactivated > 0) {
            logger.info("Grace periods expired: $deactivated user(s) deactivated")
        }

        return deactivated
    }

    private suspend fun notifyAccessRevoked(waId: String) {
        val message = """
            🔒 *Koalla — Acesso suspenso*
            
            Infelizmente não conseguimos processar o pagamento da sua assinatura e seu acesso foi suspenso.
            
            Para reativar, entre em contato com o suporte. 🐨
        """.trimIndent()

        try {
            chatwootClient.sendMessageToPhone(waId, message)
        } catch (e: Exception) {
            logger.warn("Failed to notify user $waId about suspension: ${e.message}")
        }
    }

    /**
     * Webhook SUBSCRIPTION_DELETED / SUBSCRIPTION_INACTIVATED:
     * - Marks subscription as CANCELED
     * - Deactivates user
     */
    @Transactional
    fun handleSubscriptionDeleted(subscriptionId: String) {
        val subscription = subscriptionRepository.findByAsaasSubscriptionId(subscriptionId) ?: return

        subscription.status = SubStatus.CANCELED
        subscriptionRepository.save(subscription)

        val user = userRepository.findById(subscription.userId).orElse(null)
        if (user != null) {
            user.isActive = false
            userRepository.save(user)
        }
    }

    // ── Idempotency ───────────────────────────────────────────────────────────

    fun isEventAlreadyProcessed(eventId: String): Boolean {
        return webhookEventRepository.existsByEventId(eventId)
    }

    @Transactional
    fun markEventProcessed(eventId: String, eventType: String) {
        webhookEventRepository.save(
            WebhookEvent(
                eventId = eventId,
                eventType = eventType
            )
        )
    }

    /**
     * Cancel user's active subscription.
     */
    @Transactional
    suspend fun cancelUserSubscription(user: User): CancelResult {
        val subscription = subscriptionRepository.findActiveByUserId(user.id!!)
            ?: throw IllegalStateException("User has no active subscription")

        asaasClient.cancelSubscription(subscription.asaasSubscriptionId!!)

        subscription.status = SubStatus.CANCELED
        subscriptionRepository.save(subscription)

        user.isActive = false
        userRepository.save(user)

        logger.info("Subscription canceled. user=${user.waId} subscription=${subscription.asaasSubscriptionId}")

        return CancelResult(
            success = true,
            subscriptionId = subscription.asaasSubscriptionId
        )
    }
}

data class TrialResult(
    val subscriptionId: String,
    val trialEndDate: String,
    val plan: String
)

data class CancelResult(
    val success: Boolean,
    val subscriptionId: String?
)

